src/conn/conn_compact.c

```diff
@@ -0,0 +1,260 @@
+            break;
+
+        __wt_spin_lock(session, &conn->background_compact.lock);
+        running = conn->background_compact.running;
+        if (conn->background_compact.signalled) {
+            conn->background_compact.signalled = false;
+            WT_STAT_CONN_SET(session, session_background_compact_running, running);
+        }
+        __wt_spin_unlock(session, &conn->background_compact.lock);
+
+        /*
+         * This check is necessary as we may have timed out while waiting on the mutex to be
+         * signalled and compaction is not supposed to be executed.
+         */
+        if (!running)
+            continue;
+
+        /* Open a metadata cursor. */
+        WT_ERR(__wt_metadata_cursor(session, &cursor));
+
+        cursor->set_key(cursor, key);
+        WT_ERR(cursor->search_near(cursor, &exact));
+
+        /* Make sure not to go backwards. */
+        if (exact <= 0)
+            WT_ERR_NOTFOUND_OK(cursor->next(cursor), true);
+
+        /* Find a table to compact. */
+        while (ret == 0) {
+            WT_ERR(cursor->get_key(cursor, &key));
+            /* Check we are still dealing with keys which have the right prefix. */
+            if (WT_PREFIX_MATCH(key, prefix)) {
+                /* There are files that should not be compacted. */
+                if (!WT_STREQ(key, WT_HS_URI))
+                    /* FIXME-WT-11343: check if the table is supposed to be compacted. */
+                    break;
+            } else {
+                ret = WT_NOTFOUND;
+                break;
+            }
+            WT_ERR_NOTFOUND_OK(cursor->next(cursor), true);
+        }
+
+        /*
+         * All the keys with the specified prefix have been parsed or this could be the end of the
+         * file.
+         */
+        if (ret == WT_NOTFOUND) {
+            WT_ERR(__wt_metadata_cursor_release(session, &cursor));
+            full_iteration = true;
+            continue;
+        }
+
+        /* Always close the metadata cursor. */
+        WT_ERR(__wt_metadata_cursor_release(session, &cursor));
+
+        /* Compact the file with the latest configuration. */
+        __wt_free(session, config);
+        __wt_spin_lock(session, &conn->background_compact.lock);
+        ret = __wt_strndup(session, conn->background_compact.config,
+          conn->background_compact.config == NULL ? 0 : strlen(conn->background_compact.config),
+          &config);
+        __wt_spin_unlock(session, &conn->background_compact.lock);
+
+        WT_ERR(ret);
+
+        ret = wt_session->compact(wt_session, key, config);
+
+        /* FIXME-WT-11343: compaction is done, update the data structure for this table. */
+        /*
+         * Compact may return:
+         * - EBUSY for various reasons.
+         * - ETIMEDOUT if the timer has been configured and compaction took too long.
+         * - WT_NOTFOUND if the underlying file has been deleted.
+         */
+        if (ret == EBUSY || ret == WT_NOTFOUND || ret == ETIMEDOUT)
```

Comment:
`session::compact` can also return `WT_ROLLBACK` which I think should also be okay, and we can move on to the next table.